File: app/src/lib/server/multiplayer/mouseMoves.ts
import type { BaseSocketHandler } from "./joinRoom";
import { roomExists, getPlayer } from "../rooms";
import { isSessionValid } from "./verifySession";
import type { Server } from "socket.io";

type withSessionId = (
  ...args: [
    ...BaseSocketHandler,
    io: Server,
    getSessionId: (cookie_header?: string) => string | undefined,
  ]
) => undefined;

export const mouseMoves: withSessionId = (
  socket,
  consumeRateLimit,
  io,
  getSessionId,
) => {
  socket.on("mouse_move", async (x: unknown, y: unknown) => {
    if (!(await consumeRateLimit(socket.handshake.headers.cookie ?? "", 1)))
      return;

    const roomId: unknown = socket.handshake.auth.roomId;

    if (
      typeof x !== "number" ||
      typeof y !== "number" ||
      typeof roomId !== "string"
    )
      return;

    const room_exists = await roomExists(roomId);

    if (!room_exists) {
      socket.emit("error", "Room not found");
      return;
    }

    const session_id = getSessionId(socket.handshake.headers.cookie);
    const session_valid = await isSessionValid(session_id, roomId);

    if (!session_valid) {
      socket.emit("error", "Session ID");
      return;
    }

    const player = await getPlayer(roomId, session_id!);

    if (!player) {
      socket.emit("error", "Player not found");
      return;
    }

    const { nickname, color } = player;

    io.to(`roomId/${roomId}`).emit("update_player_mouse", {
      nickname,
      color,
      x,
      y,
    });
  });
};

File: backend/src/handlers/mouse_moves.rs
use serde::Serialize;
use serde_json::json;
use socketioxide::extract::{Data, SocketRef, State};

use crate::{
    redis_client::RedisClient,
    rooms::{get_player, Player},
};

use super::{get_session_id, Auth};

#[derive(Debug, Serialize)]
struct PlayerReturn {
    nickname: String,
    color: String,
    x: f32,
    y: f32,
}

pub fn mouse_moves(socket: &SocketRef, auth: Auth) {
    socket.on(
        "mouse_move",
        |socket: SocketRef, Data::<(f32, f32)>((x, y)), State::<RedisClient>(client)| async move {
            let room_option = auth.room_id;

            if let None = room_option {
                socket.emit("error", "Room Id not found").ok();
                return;
            }

            let room_id = room_option.unwrap();

            let cookies = socket.req_parts().headers.get("cookie");
            let session_id_option = get_session_id(cookies);

            if let None = session_id_option {
                socket.emit("error", "Session Id not found").ok();
                return;
            }

            let session_id = session_id_option.unwrap();
            let player = get_player(&client, &room_id, session_id.value()).await;

            if let Err(_) = player {
                socket.emit("error", "Player not found").ok();
                return;
            }

            let Player { nickname, color } = player.unwrap();

            socket
                .within(format!("roomId/{}", room_id))
                .emit(
                    "update_player_mouse",
                    json!(PlayerReturn {
                        nickname,
                        color,
                        x,
                        y
                    }),
                )
                .ok();
        },
    );
}
